# NumPy 코드 실습 — PPT 연결본

PPT 오른쪽 위의 **N번호**로 예제를 찾으세요. N01, N02는 고정 예제 번호이며 Jupyter의 실행 횟수 `In [ ]`와 다릅니다. 각 코드 셀 첫 줄에도 PPT 페이지를 표시했습니다.

**페이지는 표지를 포함한 순서**입니다. PPT 화면 오른쪽 아래의 인쇄 쪽번호는 1 작습니다. 코드 없는 이론 페이지와 대본은 옮기지 않았습니다.

- N: 강의 코드와 변형 예제 · Q: 확인 문제 · E: 종합 실습
- `Shift+Enter`로 실행합니다. 먼저 N01을 실행하세요.
- NumPy가 없으면 새 셀에서 `%pip install numpy`를 실행하고 커널을 재시작하세요. 설치 명령은 자동 실행하지 않습니다.
- 연습문제 풀이와 자체 확인은 맨 뒤에 있습니다. 출력은 실행된 상태로 저장했습니다.


## 예제 찾기

- [N01 · PPT 7페이지 · Python을 실행할 개발 환경 준비](#n01)
- [N02 · PPT 9페이지 · 기본 배열 만들기](#n02)
- [N03 · PPT 10페이지 · 배열의 차원 구분](#n03)
- [N04 · PPT 11페이지 · 배열의 차원 구분](#n04)
- [N05 · PPT 12페이지 · 정수 범위로 배열 만들기: np.arange()](#n05)
- [N06 · PPT 13페이지 · 일정한 간격으로 배열 만들기: np.arange()](#n06)
- [N07 · PPT 14페이지 · 구간을 일정한 개수로 나누기](#n07)
- [N08 · PPT 15페이지 · 값을 초기화하지 않고 공간 확보: np.empty()](#n08)
- [N09 · PPT 16페이지 · 기존 배열을 기준으로 공간 확보: np.empty_like()](#n09)
- [N10 · PPT 17페이지 · Identity matrix(단위행렬)](#n10)
- [N11 · PPT 18페이지 · np.eye()로 단위행렬 만들기](#n11)
- [N12 · PPT 19페이지 · np.eye()의 확장 기능: 대각선 위치 변경](#n12)
- [N13 · PPT 20페이지 · 기본 배열 만들기](#n13)
- [N14 · PPT 21페이지 · 기본 배열 만들기](#n14)
- [N15 · PPT 22페이지 · np.ones(): 모든 원소를 1로 초기화](#n15)
- [N16 · PPT 23페이지 · 기존 배열과 같은 모양·자료형으로 초기화](#n16)
- [N17 · PPT 24페이지 · NumPy의 data type(자료형)](#n17)
- [N18 · PPT 26페이지 · 자료형 변환에는 조건이 필요](#n18)
- [N19 · PPT 27페이지 · 기존 배열의 자료형 바꾸기: astype()](#n19)
- [N20 · PPT 29페이지 · ndim: 배열의 차원 수 확인](#n20)
- [N21 · PPT 30페이지 · size: 배열의 전체 원소 수 확인](#n21)
- [N22 · PPT 31페이지 · shape: 각 축의 길이 확인](#n22)
- [N23 · PPT 32페이지 · reshape(): 원소를 유지하며 모양 변경](#n23)
- [N24 · PPT 33페이지 · 원하는 모양으로 모두 바꿀 수 있을까?](#n24)
- [N25 · PPT 34페이지 · Flattening(평탄화): 다차원 배열을 1차원으로](#n25)
- [N26 · PPT 36페이지 · 3차원 배열의 인덱싱](#n26)
- [N27 · PPT 37페이지 · 음수 인덱싱: 뒤에서부터 접근](#n27)
- [N28 · PPT 38페이지 · Iterating(순회): 원소를 하나씩 처리](#n28)
- [N29 · PPT 39페이지 · 다차원 배열의 순회](#n29)
- [N30 · PPT 41페이지 · Slicing(슬라이싱): 범위로 원소 선택](#n30)
- [N31 · PPT 42페이지 · 슬라이싱을 그림으로 이해하기](#n31)
- [N32 · PPT 43페이지 · 음수 인덱스·여러 축을 이용한 슬라이싱](#n32)
- [N33 · PPT 44페이지 · 간격을 지정하는 슬라이싱](#n33)
- [N34 · PPT 45페이지 · concatenate(): 기존 축을 따라 배열 연결](#n34)
- [N35 · PPT 46페이지 · stack(): 새 축을 만들어 배열 쌓기](#n35)
- [N36 · PPT 47페이지 · hstack(): 배열을 가로 방향으로 연결](#n36)
- [N37 · PPT 48페이지 · vstack(): 배열을 세로 방향으로 쌓기](#n37)
- [N38 · PPT 49페이지 · dstack(): 세 번째 축을 따라 쌓기](#n38)
- [N39 · PPT 50페이지 · 배열 연결 예제: 같은 행에 열 추가](#n39)
- [N40 · PPT 51페이지 · array_split(): 배열을 여러 부분으로 나누기](#n40)
- [N41 · PPT 52페이지 · copy(): 원본과 독립적인 배열 복사](#n41)
- [N42 · PPT 53페이지 · where(): 조건에 맞는 원소의 위치 찾기](#n42)
- [N43 · PPT 54페이지 · 여러 조건을 조합하여 검색](#n43)
- [N44 · PPT 55페이지 · intersect1d(): 두 배열의 교집합 구하기](#n44)
- [N45 · PPT 56페이지 · sort(): 배열의 값 정렬](#n45)
- [N46 · PPT 57페이지 · 정렬할 축을 선택하기](#n46)
- [N47 · PPT 58페이지 · argsort(): 정렬 순서에 해당하는 인덱스 구하기](#n47)
- [N48 · PPT 59페이지 · unique(): 중복을 제거한 값 구하기](#n48)
- [N49 · PPT 61페이지 · 기본 산술 연산: 같은 위치끼리 계산](#n49)
- [N50 · PPT 62페이지 · 기본 산술 연산: 원소별 연산](#n50)
- [N51 · PPT 63페이지 · Broadcasting(브로드캐스팅)](#n51)
- [N52 · PPT 64페이지 · 배열을 하나의 값으로 요약하기](#n52)
- [N53 · PPT 65페이지 · 자주 사용하는 수학·통계 함수](#n53)
- [N54 · PPT 66페이지 · 소수점 처리: 버림과 반올림](#n54)
- [N55 · PPT 67페이지 · 소수점 처리: 내림과 올림](#n55)
- [N56 · PPT 68페이지 · Logarithm(로그): 거듭제곱의 지수 구하기](#n56)

<a id="n01"></a>
## N01 · Python을 실행할 개발 환경 준비

PPT 7페이지 · 인쇄 쪽번호 6

In [3]:
# N01 · PPT 7페이지 (인쇄 쪽번호 6) · 기본 예제
import numpy as np
print('NumPy 버전:', np.__version__)

NumPy 버전: 2.0.2


In [3]:
# N01 · PPT 7페이지 (인쇄 쪽번호 6) · 변형 예제
import numpy
print(numpy.__version__)
print(np is numpy)
print(np.__version__)

2.1.3
True
2.1.3


as np는 같은 모듈을 짧은 이름으로 부르게 합니다. True는 두 이름이 같은 모듈을 가리킨다는 뜻입니다.

<a id="n02"></a>
## N02 · 기본 배열 만들기

PPT 9페이지 · 인쇄 쪽번호 8

In [4]:
# N02 · PPT 9페이지 (인쇄 쪽번호 8) · 기본 예제
arr = np.array([1, 2, 3])
print(arr)

[1 2 3]


**바꿔 보면? 값 하나를 추가하거나 실수를 섞으면?**

In [5]:
# N02 · PPT 9페이지 (인쇄 쪽번호 8) · 변형 예제
print(np.array([1,2,3,4]))
print(np.array([1,2,3.5]).dtype)

[1 2 3 4]
float64


원소가 하나 늘면 배열 길이가 4가 됩니다. 실수가 섞이면 정수들도 함께 표현할 수 있는 부동소수점 자료형을 선택합니다.

<a id="n03"></a>
## N03 · 배열의 차원 구분

PPT 10페이지 · 인쇄 쪽번호 9

In [6]:
# N03 · PPT 10페이지 (인쇄 쪽번호 9) · 기본 예제
arr = np.array(42)
print('0차원:', arr, 'shape:', arr.shape)
arr = np.array([1, 2, 3, 4, 5])
print('1차원:', arr, 'shape:', arr.shape)

0차원: 42 shape: ()
1차원: [1 2 3 4 5] shape: (5,)


**바꿔 보면? 42와 [42]는 같은 배열일까?**

In [10]:
# N03 · PPT 10페이지 (인쇄 쪽번호 9) · 변형 예제
for a in (np.array(42), np.array([42])):
    print(a, "ndim:", a.ndim, "shape:", a.shape)

42 ndim: 0 shape: ()
[42] ndim: 1 shape: (1,)


값이 같아도 42는 0차원, [42]는 길이 1인 1차원입니다.

<a id="n04"></a>
## N04 · 배열의 차원 구분

PPT 11페이지 · 인쇄 쪽번호 10

In [15]:
# N04 · PPT 11페이지 (인쇄 쪽번호 10) · 기본 예제
arr = np.array([[1, 2, 3], [4, 5, 6]])
print('2차원:', arr, 'shape:', arr.shape)
print('\n')

arr = np.array([[[1, 2, 3], [4, 5, 6]], [[1, 2, 3], [4, 5, 6]]])
print('3차원:', arr, 'shape:', arr.shape)

2차원: [[1 2 3]
 [4 5 6]] shape: (2, 3)


3차원: [[[1 2 3]
  [4 5 6]]

 [[1 2 3]
  [4 5 6]]] shape: (2, 2, 3)


**바꿔 보면? 대괄호를 한 겹 더 쓰면?**

In [ ]:
# N04 · PPT 11페이지 (인쇄 쪽번호 10) · 변형 예제
a=np.array([[1,2,3],[4,5,6]])
print(a.shape)
print(np.array([a]).shape)

(2, 3)
(1, 2, 3)


(2, 3)에 길이 1인 축이 앞에 추가되어 (1, 2, 3)이 됩니다.

<a id="n05"></a>
## N05 · 정수 범위로 배열 만들기: np.arange()

PPT 12페이지 · 인쇄 쪽번호 11

In [ ]:
# N05 · PPT 12페이지 (인쇄 쪽번호 11) · 기본 예제
print(np.arange(4))
print('슬라이드 그림의 별도 예:', np.arange(5))

[0 1 2 3]
슬라이드 그림의 별도 예: [0 1 2 3 4]


**바꿔 보면? 끝 경계가 1 커지거나 0이면?**

In [19]:
# N05 · PPT 12페이지 (인쇄 쪽번호 11) · 변형 예제

print(np.arange(1))
print(np.arange(0))
print(np.arange(5))
print(np.arange(10))

[0]
int64
[0 1 2 3 4]
[0 1 2 3 4 5 6 7 8 9]


끝 경계는 제외합니다. arange(0)은 범위 안에 원소가 없어 빈 배열입니다.

<a id="n06"></a>
## N06 · 일정한 간격으로 배열 만들기: np.arange()

PPT 13페이지 · 인쇄 쪽번호 12

In [20]:
# N06 · PPT 13페이지 (인쇄 쪽번호 12) · 기본 예제
print(np.arange(2, 9, 2))
print(np.arange(5, 9))

[2 4 6 8]
[5 6 7 8]


**바꿔 보면? 끝값까지 넣거나 역순으로 만들려면?**

In [26]:
# N06 · PPT 13페이지 (인쇄 쪽번호 12) · 변형 예제
print(np.arange(2,3,0.1))
print(np.arange(8,0,-2))

[2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9]
[8 6 4 2]


첫 결과는 여전히 [2,4,6,8]입니다. 끝값 10은 빠집니다. 음수 간격은 감소하는 범위를 만듭니다.

<a id="n07"></a>
## N07 · 구간을 일정한 개수로 나누기

PPT 14페이지 · 인쇄 쪽번호 13

In [29]:
# N07 · PPT 14페이지 (인쇄 쪽번호 13) · 기본 예제
print('linspace:', np.linspace(0, 10, num=5))


linspace: [ 0.   2.5  5.   7.5 10. ]


**바꿔 보면? num=4로 줄이면? 끝값을 제외하면?**

In [30]:
# N07 · PPT 14페이지 (인쇄 쪽번호 13) · 변형 예제
print("num=5:", np.linspace(0,10,num=5))
print("num=4:", np.linspace(0,10,num=4))
print("끝값 제외:", np.linspace(0,10,num=4,endpoint=False))

num=5: [ 0.   2.5  5.   7.5 10. ]
num=4: [ 0.          3.33333333  6.66666667 10.        ]
끝값 제외: [0.  2.5 5.  7.5]


num은 간격이 아니라 값의 개수입니다. num=4는 양 끝을 포함한 4개 값 [0, 3.333333…, 6.666667…, 10]입니다. 간격은 10/(4-1)입니다. endpoint=False이면 10을 제외하여 [0, 2.5, 5, 7.5]가 됩니다.

같은 페이지에 소개된 다른 범위·격자 생성 함수

In [36]:
# N07 · PPT 14페이지 (인쇄 쪽번호 13) · 추가 함수
# 보충: 슬라이드에서 이름으로 소개한 함수들도 실행해 봅니다.
print('logspace:', np.logspace(0, 3, num=4))  # 10의 0, 1, 2, 3제곱
print('geomspace:', np.geomspace(1, 8, num=4))  # 1, 2, 4, 8

logspace: [   1.   10.  100. 1000.]
geomspace: [1. 2. 4. 8.]


In [49]:
x = np.array([0, 1, 2])
y = np.array([10, 20])
X, Y = np.meshgrid(x, y)
print('meshgrid X:', X, 'meshgrid Y:', Y, sep='\n')
print('\n mgrid:', np.mgrid[0:2, 0:3], sep='\n')
row, col = np.ogrid[0:2, 0:3]
print('\n ogrid의 축별 shape:', row.shape, col.shape)
print('\n 열린 격자의 브로드캐스팅 결과:', row + col, sep='\n')

meshgrid X:
[[0 1 2]
 [0 1 2]]
meshgrid Y:
[[10 10 10]
 [20 20 20]]

 mgrid:
[[[0 0 0]
  [1 1 1]]

 [[0 1 2]
  [0 1 2]]]

 ogrid의 축별 shape: (2, 1) (1, 3)

 열린 격자의 브로드캐스팅 결과:
[[0 1 2]
 [1 2 3]]


<a id="n08"></a>
## N08 · 값을 초기화하지 않고 공간 확보: np.empty()

PPT 15페이지 · 인쇄 쪽번호 14

In [79]:
arr = np.empty(3)
print(arr)

[99. 99. 99.]


**바꿔 보면? shape와 dtype을 바꿔도 0이 들어갈까?**

In [62]:
# N08 · PPT 15페이지 (인쇄 쪽번호 14) · 변형 예제
a=np.empty((2,3),dtype=int)
print(a.shape, a.dtype)
a[:]=7
print(a)

(2, 3) int64
[[7 7 7]
 [7 7 7]]


empty는 모양과 자료형만 정합니다. 초기 값은 보장하지 않으므로 모든 칸을 채운 뒤 계산에 사용합니다.

<a id="n09"></a>
## N09 · 기존 배열을 기준으로 공간 확보: np.empty_like()

PPT 16페이지 · 인쇄 쪽번호 15

In [81]:
# N09 · PPT 16페이지 (인쇄 쪽번호 15) · 기본 예제
arr = np.array([[1, 2, 3], [4, 5, 6]])
print(arr.shape)
result = np.empty_like(arr)
print('초기화 전:', result)
result[:] = arr * 2
print('모든 칸을 계산 결과로 채운 뒤:', result)

(2, 3)
초기화 전: [[1 2 3]
 [4 5 6]]
모든 칸을 계산 결과로 채운 뒤: [[ 2  4  6]
 [ 8 10 12]]


**바꿔 보면? 정수 배열을 기준으로 실수 결과 공간을 만들려면?**

In [84]:
# N09 · PPT 16페이지 (인쇄 쪽번호 15) · 변형 예제
a=np.array([1,2,3])
b=np.empty_like(a,dtype=float)
print(b)
b[:]= a/2
print(b,b.dtype)

[4.9e-324 9.9e-324 1.5e-323]
[0.5 1.  1.5] float64


dtype을 지정하면 원본의 정수 자료형 대신 실수 자료형을 사용합니다.

<a id="n10"></a>
## N10 · Identity matrix(단위행렬)

PPT 17페이지 · 인쇄 쪽번호 16

In [ ]:
# N10 · PPT 17페이지 (인쇄 쪽번호 16) · 기본 예제
I = np.identity(3)
print('3×3 단위행렬:', I)
A = np.array([2, 3, 4])
print('A @ I:', A @ I)
np.testing.assert_array_equal(A @ I, A)
# 보충: @는 행렬 곱, *는 원소별 곱입니다.
A = np.array([[2, 3, 4], [5, 6, 7], [8, 9, 10]])
print('행렬 곱 A @ I:', A @ I)
print('원소별 곱 A * I:', A * I)

3×3 단위행렬: [[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
A @ I: [2. 3. 4.]
행렬 곱 A @ I: [[ 2.  3.  4.]
 [ 5.  6.  7.]
 [ 8.  9. 10.]]
원소별 곱 A * I: [[ 2.  0.  0.]
 [ 0.  6.  0.]
 [ 0.  0. 10.]]


**바꿔 보면? 단위행렬 크기를 바꾸면? *로 곱하면?**

In [ ]:
# N10 · PPT 17페이지 (인쇄 쪽번호 16) · 변형 예제
A=np.array([[2,3],[4,5]])
I=np.identity(2)
print("@:", A@I)
print("*:", A*I)

@: [[2. 3.]
 [4. 5.]]
*: [[2. 0.]
 [0. 5.]]


@는 원래 행렬을 보존합니다. *는 같은 위치끼리 곱하므로 대각선 밖 원소가 0이 됩니다.

<a id="n11"></a>
## N11 · np.eye()로 단위행렬 만들기

PPT 18페이지 · 인쇄 쪽번호 17

In [85]:
# N11 · PPT 18페이지 (인쇄 쪽번호 17) · 기본 예제
print(np.eye(4))

[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


**바꿔 보면? 행과 열의 개수를 다르게 지정하면?**

In [93]:
# N11 · PPT 18페이지 (인쇄 쪽번호 17) · 변형 예제
print(np.eye(2,4))
print(np.eye(2,4).dtype)

print(np.eye(2, 4, k=1))
print(np.eye(3,dtype=int))

[[1. 0. 0. 0.]
 [0. 1. 0. 0.]]
float64
[[0. 1. 0. 0.]
 [0. 0. 1. 0.]]
[[1 0 0]
 [0 1 0]
 [0 0 1]]


eye(2,4)는 2행 4열입니다. 정사각형이 아니므로 단위행렬이라고 부르지 않습니다. dtype=int는 1과 0을 정수로 저장합니다.

<a id="n12"></a>
## N12 · np.eye()의 확장 기능: 대각선 위치 변경

PPT 19페이지 · 인쇄 쪽번호 18

In [ ]:
# N12 · PPT 19페이지 (인쇄 쪽번호 18) · 기본 예제
print(np.eye(5, k=1))
print('보충: 직사각형 배열:', np.eye(2, 4))

[[0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]
 [0. 0. 0. 0. 0.]]
보충: 직사각형 배열: [[1. 0. 0. 0.]
 [0. 1. 0. 0.]]


**바꿔 보면? k=-1 또는 k=0이면?**

In [ ]:
# N12 · PPT 19페이지 (인쇄 쪽번호 18) · 변형 예제
print(np.eye(4,k=-1))
print(np.eye(4,k=0))

[[0. 0. 0. 0.]
 [1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]]
[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


음수 k는 주대각선 아래, 0은 주대각선입니다.

<a id="n13"></a>
## N13 · 기본 배열 만들기

PPT 20페이지 · 인쇄 쪽번호 19

In [ ]:
# N13 · PPT 20페이지 (인쇄 쪽번호 19) · 기본 예제
print(np.zeros(2))
print('슬라이드 그림의 별도 예:', np.zeros(6))

[0. 0.]
슬라이드 그림의 별도 예: [0. 0. 0. 0. 0. 0.]


**바꿔 보면? 0 두 개와 0 여섯 개는? 정수 0은?**

In [ ]:
# N13 · PPT 20페이지 (인쇄 쪽번호 19) · 변형 예제
print(np.zeros(6))
print(np.zeros(2,dtype=int))

[0. 0. 0. 0. 0. 0.]
[0 0]


첫 인자는 원소 수입니다. dtype을 생략하면 기본적으로 부동소수점 0.0으로 채웁니다.

<a id="n14"></a>
## N14 · 기본 배열 만들기

PPT 21페이지 · 인쇄 쪽번호 20

In [ ]:
# N14 · PPT 21페이지 (인쇄 쪽번호 20) · 기본 예제
a = (3, 2)
print(np.zeros(a))

[[0. 0.]
 [0. 0.]
 [0. 0.]]


**바꿔 보면? (3,2)를 (2,3)으로 바꾸면?**

In [ ]:
# N14 · PPT 21페이지 (인쇄 쪽번호 20) · 변형 예제
print(np.zeros((2,3)))
print(np.zeros((1,2,3)).shape)

[[0. 0. 0.]
 [0. 0. 0.]]
(1, 2, 3)


원소 수 6은 같지만 행·열 배치가 바뀝니다. (1,2,3)은 길이 1인 축까지 가진 3차원입니다.

<a id="n15"></a>
## N15 · np.ones(): 모든 원소를 1로 초기화

PPT 22페이지 · 인쇄 쪽번호 21

In [94]:
# N15 · PPT 22페이지 (인쇄 쪽번호 21) · 기본 예제
print(np.ones(2))
print(np.full(3, 7))
print(np.full(shape=3, fill_value=7))

[1. 1.]
[7 7 7]
[7 7 7]


**바꿔 보면? full에 정수 대신 실수를 넣으면?**

In [95]:
# N15 · PPT 22페이지 (인쇄 쪽번호 21) · 변형 예제
print(np.full((2,3),7))
a=np.full(3,7.5)
print(a,a.dtype)

[[7 7 7]
 [7 7 7]]
[7.5 7.5 7.5] float64


shape는 튜플도 받을 수 있습니다. dtype을 생략한 full은 채울 값에서 자료형을 추론합니다.

<a id="n16"></a>
## N16 · 기존 배열과 같은 모양·자료형으로 초기화

PPT 23페이지 · 인쇄 쪽번호 22

In [101]:
# N16 · PPT 23페이지 (인쇄 쪽번호 22) · 기본 예제
a = np.array([[2, 3], [4, 5]])
print('ones_like:', np.ones_like(a))
print('zeros_like:', np.zeros_like(a))
print('full_like:', np.full_like(a, 9))

ones_like: [[1 1]
 [1 1]]
zeros_like: [[0 0]
 [0 0]]
full_like: [[9 9]
 [9 9]]


**바꿔 보면? 정수 배열에 full_like로 1.5를 채우면?**

In [ ]:
# N16 · PPT 23페이지 (인쇄 쪽번호 22) · 변형 예제
a=np.array([1,2,3])
print(np.full_like(a,1.5))
print(np.full_like(a,1.5,dtype=float))

[1 1 1]
[1.5 1.5 1.5]


기본값은 원본의 자료형을 따르므로 1.5가 정수 1로 바뀝니다. 실수를 유지하려면 dtype=float를 지정합니다.

<a id="n17"></a>
## N17 · NumPy의 data type(자료형)

PPT 24페이지 · 인쇄 쪽번호 23

In [104]:
# N17 · PPT 24페이지 (인쇄 쪽번호 23) · 기본 예제
arr = np.array([0, 1, 2, 3])
print('기본 정수 자료형:', arr.dtype)  # 환경에 따라 int32 또는 int64 등
arr = np.array([1, 2, 3, 4], dtype='f')

print(arr)
print(arr.dtype)

기본 정수 자료형: int64
[1. 2. 3. 4.]
float32


**바꿔 보면? 자료형을 int32 또는 float64로 지정하면?**

In [106]:
# N17 · PPT 24페이지 (인쇄 쪽번호 23) · 변형 예제
for t in (np.int32,np.float64):
    a=np.array([1,2,3],dtype=t)
    print(a,a.dtype,a.itemsize)

[1 2 3] int32 4
[1. 2. 3.] float64 8


자료형을 명시하면 저장 방식이 명확해집니다. itemsize는 원소 하나가 차지하는 바이트 수입니다.

In [111]:
a = np.array([1, 2, 3], dtype=np.int32)
b = np.array([1, 2, 3], dtype=np.int64)

print(a.dtype)       # int32
print(b.dtype)       # int64

print(a.dtype.kind)  # i
print(b.dtype.kind)  # i

int32
int64
i
i


In [110]:
a = np.array([True, False], dtype=bool)
print(a.dtype.kind)  # b: 불리언 계열이라는 뜻

b = np.array([1, 2], dtype='b')
print(b.dtype)       # int8: 여기서 'b'는 8비트 정수를 지정

b
int8


<a id="n18"></a>
## N18 · 자료형 변환에는 조건이 필요

PPT 26페이지 · 인쇄 쪽번호 25

In [112]:
# N18 · PPT 26페이지 (인쇄 쪽번호 25) · 기본 예제
# 의도된 오류: 예외를 잡아 뒤의 셀도 계속 실행되게 합니다.
try:
    arr = np.array(['a', '2', '3'], dtype='i')
    print(arr)
    print(arr.dtype)
except ValueError as error:
    print('예상한 ValueError:', error)
print('변환 가능한 문자열:', np.array(['1', '2', '3'], dtype='i'))

예상한 ValueError: invalid literal for int() with base 10: 'a'
변환 가능한 문자열: [1 2 3]


**바꿔 보면? 문자열 3.5를 곧바로 정수로 바꿀 수 있을까?**

In [113]:
# N18 · PPT 26페이지 (인쇄 쪽번호 25) · 변형 예제
for t in (int,float):
    try:
        print(np.array(["3.5","2"],dtype=t))
    except ValueError as e:
        print("예상한 오류:",e)

예상한 오류: invalid literal for int() with base 10: '3.5'
[3.5 2. ]


3.5는 정수 형식 문자열이 아니므로 int 변환은 실패합니다. float 변환은 가능합니다.

<a id="n19"></a>
## N19 · 기존 배열의 자료형 바꾸기: astype()

PPT 27페이지 · 인쇄 쪽번호 26

In [114]:
# N19 · PPT 27페이지 (인쇄 쪽번호 26) · 기본 예제
arr = np.array([1.1, 2.1, 3.1])
newarr = arr.astype('i')
print(newarr)
print(newarr.dtype)
print('원본:', arr)
print('음수의 변환은 내림과 다름:', np.array([-3.9]).astype(int))

[1 2 3]
int32
원본: [1.1 2.1 3.1]
음수의 변환은 내림과 다름: [-3]


**바꿔 보면? 음수 소수를 정수로 바꾸면 내림일까?**

In [ ]:
# N19 · PPT 27페이지 (인쇄 쪽번호 26) · 변형 예제
a=np.array([-3.9,3.9])
print("astype:",a.astype(int))
print("floor:",np.floor(a))

astype:

 [-3  3]
floor: [-4.  3.]


astype(int)는 소수부를 0 방향으로 버립니다. -3.9는 -3이지만 floor는 -4입니다.

<a id="n20"></a>
## N20 · ndim: 배열의 차원 수 확인

PPT 29페이지 · 인쇄 쪽번호 28

In [125]:
# N20 · PPT 29페이지 (인쇄 쪽번호 28) · 기본 예제
array_example = np.array([[[0, 1, 2, 3], [4, 5, 6, 7]],
                          [[0, 1, 2, 3], [4, 5, 6, 7]],
                          [[0, 1, 2, 3], [4, 5, 6, 7]]])
print(array_example.ndim)

3


**바꿔 보면? 축 길이가 1인 경우에도 차원이 늘어날까?**

In [120]:
# N20 · PPT 29페이지 (인쇄 쪽번호 28) · 변형 예제
print(np.array([1,2,3]).ndim)
print(np.array([[1,2,3]]).ndim)

1
2


행이 하나여도 대괄호가 구조를 한 겹 더 만들면 2차원입니다.

<a id="n21"></a>
## N21 · size: 배열의 전체 원소 수 확인

PPT 30페이지 · 인쇄 쪽번호 29

In [ ]:
# N21 · PPT 30페이지 (인쇄 쪽번호 29) · 기본 예제
array_example = np.array([[[0, 1, 2, 3], [4, 5, 6, 7]],
                          [[0, 1, 2, 3], [4, 5, 6, 7]],
                          [[0, 1, 2, 3], [4, 5, 6, 7]]])
print(array_example.size)

24


**바꿔 보면? 모양만 바꾸면 size도 변할까?**

In [127]:
# N21 · PPT 30페이지 (인쇄 쪽번호 29) · 변형 예제
a=np.arange(12)
print(a)
print(a.size,a.reshape(3,4).size,a.reshape(2,2,3).size)

[ 0  1  2  3  4  5  6  7  8  9 10 11]
12 12 12


모양을 바꾸어도 원소 12개는 유지됩니다.

<a id="n22"></a>
## N22 · shape: 각 축의 길이 확인

PPT 31페이지 · 인쇄 쪽번호 30

In [128]:
# N22 · PPT 31페이지 (인쇄 쪽번호 30) · 기본 예제
array_example = np.array([[[0, 1, 2, 3], [4, 5, 6, 7]],
                          [[0, 1, 2, 3], [4, 5, 6, 7]],
                          [[0, 1, 2, 3], [4, 5, 6, 7]]])
print(array_example.shape)

(3, 2, 4)


**바꿔 보면? (3,)과 (1,3)은 어떻게 다를까?**

In [129]:
# N22 · PPT 31페이지 (인쇄 쪽번호 30) · 변형 예제
for a in (np.array([1,2,3]),np.array([[1,2,3]]),np.array([[1],[2],[3]])):
    print(a.shape)

(3,)
(1, 3)
(3, 1)


np.array([1,2,3])
“1, 2, 3을 원소로 가지는 1차원 배열” — shape: (3,)

np.array([[1,2,3]])
“한 행에 1, 2, 3이 들어 있는 1행 3열의 2차원 배열” — shape: (1, 3)

np.array([[1],[2],[3]])
“각 행에 1, 2, 3이 하나씩 들어 있는 3행 1열의 2차원 배열” — shape: (3, 1)

(3,)은 한 축, (1,3)과 (3,1)은 두 축입니다. 축 개수와 각 축의 길이를 구분합니다.

<a id="n23"></a>
## N23 · reshape(): 원소를 유지하며 모양 변경

PPT 32페이지 · 인쇄 쪽번호 31

In [133]:
# N23 · PPT 32페이지 (인쇄 쪽번호 31) · 기본 예제
a = np.arange(6)
print(a)
b = a.reshape(3, 2)
print(b)

[0 1 2 3 4 5]
[[0 1]
 [2 3]
 [4 5]]


**바꿔 보면? 같은 6개를 2행으로 바꾸거나 -1을 쓰면?**

In [134]:
# N23 · PPT 32페이지 (인쇄 쪽번호 31) · 변형 예제
a=np.arange(6)
print(a)
print(a.reshape(2,3))
print(a.reshape(3,-1))

[0 1 2 3 4 5]
[[0 1 2]
 [3 4 5]]
[[0 1]
 [2 3]
 [4 5]]


2행으로 정하면 열 수는 6/2=3입니다. -1이 이를 계산합니다.

<a id="n24"></a>
## N24 · 원하는 모양으로 모두 바꿀 수 있을까?

PPT 33페이지 · 인쇄 쪽번호 32

In [135]:
# N24 · PPT 33페이지 (인쇄 쪽번호 32) · 기본 예제
arr = np.array([1, 2, 3, 4, 5, 6, 7, 8])
try:
    newarr = arr.reshape(3, 3)
    print(newarr)
except ValueError as error:
    print('예상한 ValueError:', error)

예상한 ValueError: cannot reshape array of size 8 into shape (3,3)


**바꿔 보면? 8개를 (2,4) 또는 (2,-1)로 바꾸면?**

In [137]:
# N24 · PPT 33페이지 (인쇄 쪽번호 32) · 변형 예제
a=np.arange(1,9)
print(a)
print(a.reshape(2,4))
print(a.reshape(2,-1))

[1 2 3 4 5 6 7 8]
[[1 2 3 4]
 [5 6 7 8]]
[[1 2 3 4]
 [5 6 7 8]]


둘 다 원소 수가 8개이므로 가능합니다.

<a id="n25"></a>
## N25 · Flattening(평탄화): 다차원 배열을 1차원으로

PPT 34페이지 · 인쇄 쪽번호 33

In [143]:
# N25 · PPT 34페이지 (인쇄 쪽번호 33) · 기본 예제
arr = np.array([[1, 2, 3], [4, 5, 6]])
newarr = arr.reshape(-1)
print(newarr)

[1 2 3 4 5 6]


**바꿔 보면? 1차원 대신 열벡터 모양으로 펴려면?**

In [144]:
# N25 · PPT 34페이지 (인쇄 쪽번호 33) · 변형 예제
a=np.array([[1,2,3],[4,5,6]])
print(a.reshape(-1))
print(a.reshape(-1,1))

[1 2 3 4 5 6]
[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]


(-1)은 (6,)이고 (-1,1)은 (6,1)입니다. 두 번째는 열이 하나인 2차원입니다.

<a id="q01"></a>
### Q01 · 확인1 · PPT 34페이지

[1, 2, ..., 12]를 shape (2, 2, 3)인 3차원 배열로 바꾸세요. 원소 수가 같은 다른 3차원 모양도 가능하지만 이 실습에서는 (2, 2, 3)을 사용합니다.

`answer`에 결과를 저장하세요. [풀이](#q01-solution)

In [149]:
# Q01 · PPT 34페이지 (인쇄 쪽번호 33) · 직접 풀기
a = np.arange(1, 13)
answer = a.reshape(2,2,3)
print(answer, '\n', answer.ndim)
# 여기에 풀이를 작성하세요.

[[[ 1  2  3]
  [ 4  5  6]]

 [[ 7  8  9]
  [10 11 12]]] 
 3


<a id="n26"></a>
## N26 · 3차원 배열의 인덱싱

PPT 36페이지 · 인쇄 쪽번호 35

In [151]:
# N26 · PPT 36페이지 (인쇄 쪽번호 35) · 기본 예제
arr = np.array([[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]])
print(arr.shape)
print(arr[0, 1, 2])

(2, 2, 3)
6


**바꿔 보면? 묶음 또는 행 인덱스를 바꾸면?**

In [152]:
# N26 · PPT 36페이지 (인쇄 쪽번호 35) · 변형 예제
a=np.arange(1,13).reshape(2,2,3)
print(a[1,0,2])
print(a[0,1,:])

9
[4 5 6]


첫 결과는 두 번째 묶음의 첫 행 셋째 값 9입니다. :는 해당 행의 모든 열 [4,5,6]을 선택합니다.

<a id="n27"></a>
## N27 · 음수 인덱싱: 뒤에서부터 접근

PPT 37페이지 · 인쇄 쪽번호 36

In [154]:
# N27 · PPT 37페이지 (인쇄 쪽번호 36) · 기본 예제
arr = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10]])
print(arr.shape)
print(arr[1, -1])

(2, 5)
10


**바꿔 보면? -1,-1과 0,-2는?**

In [ ]:
# N27 · PPT 37페이지 (인쇄 쪽번호 36) · 변형 예제
a=np.arange(1,11).reshape(2,5)
print(a[-1,-1])
print(a[0,-2])

10
4


마지막 행 마지막 값은 10, 첫 행 뒤에서 둘째 값은 4입니다.

<a id="n28"></a>
## N28 · Iterating(순회): 원소를 하나씩 처리

PPT 38페이지 · 인쇄 쪽번호 37

In [156]:
# N28 · PPT 38페이지 (인쇄 쪽번호 37) · 기본 예제
arr = np.array([1, 2, 3])
for x in arr:
    print(x)

1
2
3


**바꿔 보면? 반복 대신 배열 전체를 두 배 하면?**

In [155]:
# N28 · PPT 38페이지 (인쇄 쪽번호 37) · 변형 예제
a=np.array([1,2,3])
print([x*2 for x in a])
print(a*2)

[np.int64(2), np.int64(4), np.int64(6)]
[2 4 6]


값은 같지만 첫 결과는 Python 리스트, 두 번째는 NumPy 배열입니다.

<a id="n29"></a>
## N29 · 다차원 배열의 순회

PPT 39페이지 · 인쇄 쪽번호 38

In [158]:
# N29 · PPT 39페이지 (인쇄 쪽번호 38) · 기본 예제
arr = np.array([[1, 2, 3], [4, 5, 6]])
for x in arr:
    print(x)
    for y in x:
        print(y)

[1 2 3]
1
2
3
[4 5 6]
4
5
6


**바꿔 보면? 2차원 배열에서 for를 한 번만 쓰면?**

In [ ]:
# N29 · PPT 39페이지 (인쇄 쪽번호 38) · 변형 예제
a=np.array([[1,2,3],[4,5,6]])
for row in a:
    print("이번 반복에서 받은 행:",row)

이번 반복에서 받은 행: [1 2 3]
이번 반복에서 받은 행: [4 5 6]


첫 축을 따라 이동하므로 한 번에 행 배열 하나를 받습니다.

<a id="n30"></a>
## N30 · Slicing(슬라이싱): 범위로 원소 선택

PPT 41페이지 · 인쇄 쪽번호 40

In [163]:
# N30 · PPT 41페이지 (인쇄 쪽번호 40) · 기본 예제
data = np.array([1, 2, 3])
print(data[1])
print(data[0:2])
print(data[1:])
print(data[-2:])

2
[1 2]
[2 3]
[2 3]


**바꿔 보면? 끝 인덱스를 크게 쓰거나 같은 시작·끝을 쓰면?**

In [164]:
# N30 · PPT 41페이지 (인쇄 쪽번호 40) · 변형 예제
a=np.array([1,2,3])
print(a[:10])
print(a[1:1])

[1 2 3]
[]


슬라이스의 끝이 길이를 넘어도 가능한 범위까지만 선택합니다. 시작과 끝이 같으면 빈 배열입니다.

<a id="n31"></a>
## N31 · 슬라이싱을 그림으로 이해하기

PPT 42페이지 · 인쇄 쪽번호 41

In [ ]:
# N31 · PPT 42페이지 (인쇄 쪽번호 41) · 기본 예제
data = np.array([1, 2, 3])
print(data[1])
print(data[0:2])
print(data[1:])
print(data[-2:])

2
[1 2]
[2 3]
[2 3]


**바꿔 보면? 슬라이스의 값을 바꾸면 원본도 바뀔까?**

In [170]:
# N31 · PPT 42페이지 (인쇄 쪽번호 41) · 변형 예제
a=np.array([1,2,3])
b = a[1:]
print(b)
b[0] = 99
print(b)
print("원본:",a)

a = np.array([1,2,3])
b=a[1:].copy()
b[0]=99
print("copy 사용 시 원본:", a)

[2 3]
[99  3]
원본: [ 1 99  3]
copy 사용 시 원본: [1 2 3]


기본 슬라이싱은 원본과 데이터를 공유하는 뷰입니다. 독립적으로 수정하려면 copy를 사용합니다.

<a id="n32"></a>
## N32 · 음수 인덱스·여러 축을 이용한 슬라이싱

PPT 43페이지 · 인쇄 쪽번호 42

In [171]:
# N32 · PPT 43페이지 (인쇄 쪽번호 42) · 기본 예제
arr = np.array([1, 2, 3, 4, 5, 6, 7])
print(arr[-3:-1])
arr = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10]])
print(arr[0:2, 2])

[5 6]
[3 8]


<a id="n33"></a>
## N33 · 간격을 지정하는 슬라이싱

PPT 44페이지 · 인쇄 쪽번호 43

In [ ]:
# N33 · PPT 44페이지 (인쇄 쪽번호 43) · 기본 예제
arr = np.array([1, 2, 3, 4, 5, 6, 7])
print(arr[1:5:2])
print(arr[::2])

[2 4]
[1 3 5 7]


**바꿔 보면? 간격을 3 또는 -1로 바꾸면?**

In [174]:
# N33 · PPT 44페이지 (인쇄 쪽번호 43) · 변형 예제
a=np.arange(1,8)
print(a)
print(a[::3])
print(a[::-1])

[1 2 3 4 5 6 7]
[1 4 7]
[7 6 5 4 3 2 1]


step=3은 세 칸 간격으로 [1,4,7]을 읽고 -1은 전체를 역순으로 읽습니다.

<a id="n34"></a>
## N34 · concatenate(): 기존 축을 따라 배열 연결

PPT 45페이지 · 인쇄 쪽번호 44

In [175]:
# N34 · PPT 45페이지 (인쇄 쪽번호 44) · 기본 예제
arr1 = np.array([1, 2, 3])
arr2 = np.array([4, 5, 6])
arr = np.concatenate((arr1, arr2))
print(arr)
print("shape:", arr.shape)

[1 2 3 4 5 6]
shape: (6,)


**바꿔 보면? 2차원 배열에서 axis=0과 1은?**

In [178]:
# N34 · PPT 45페이지 (인쇄 쪽번호 44) · 변형 예제
a=np.array([[1,2],[3,4]])
b=np.array([[5,6],[7,8]])
print(np.concatenate((a,b),axis=0))
print(np.concatenate((a,b),axis=0).shape)
print(np.concatenate((a,b),axis=1))
print(np.concatenate((a,b),axis=1).shape)

[[1 2]
 [3 4]
 [5 6]
 [7 8]]
(4, 2)
[[1 2 5 6]
 [3 4 7 8]]
(2, 4)


axis=0은 행 방향으로 이어 (4,2), axis=1은 열 방향으로 이어 (2,4)가 됩니다.

<a id="n35"></a>
## N35 · stack(): 새 축을 만들어 배열 쌓기

PPT 46페이지 · 인쇄 쪽번호 45

In [179]:
# N35 · PPT 46페이지 (인쇄 쪽번호 45) · 기본 예제
arr1 = np.array([1, 2, 3])
arr2 = np.array([4, 5, 6])
arr = np.stack((arr1, arr2), axis=1)
print(arr)
print("shape:", arr.shape)

[[1 4]
 [2 5]
 [3 6]]
shape: (3, 2)


**바꿔 보면? stack의 axis=0과 1은?**

In [180]:
# N35 · PPT 46페이지 (인쇄 쪽번호 45) · 변형 예제
a=np.array([1,2,3]); b=np.array([4,5,6])
print(np.stack((a,b),axis=0))
print(np.stack((a,b),axis=1))

[[1 2 3]
 [4 5 6]]
[[1 4]
 [2 5]
 [3 6]]


새 축의 위치가 달라져 (2,3)과 (3,2)가 됩니다.

<a id="n36"></a>
## N36 · hstack(): 배열을 가로 방향으로 연결

PPT 47페이지 · 인쇄 쪽번호 46

In [181]:
# N36 · PPT 47페이지 (인쇄 쪽번호 46) · 기본 예제
arr1 = np.array([1, 2, 3])
arr2 = np.array([4, 5, 6])
arr = np.hstack((arr1, arr2))
print(arr)
print("shape:", arr.shape)

[1 2 3 4 5 6]
shape: (6,)


**바꿔 보면? 2차원 배열을 hstack하면?**

In [182]:
# N36 · PPT 47페이지 (인쇄 쪽번호 46) · 변형 예제
a=np.array([[1],[2]])
b=np.array([[3],[4]])
print(np.hstack((a,b)))

[[1 3]
 [2 4]]


두 배열의 행 수를 유지하고 열을 이어 [[1,3],[2,4]]로 만듭니다.

<a id="n37"></a>
## N37 · vstack(): 배열을 세로 방향으로 쌓기

PPT 48페이지 · 인쇄 쪽번호 47

In [183]:
# N37 · PPT 48페이지 (인쇄 쪽번호 47) · 기본 예제
arr1 = np.array([1, 2, 3])
arr2 = np.array([4, 5, 6])
arr = np.vstack((arr1, arr2))
print(arr)
print("shape:", arr.shape)

[[1 2 3]
 [4 5 6]]
shape: (2, 3)


<a id="n38"></a>
## N38 · dstack(): 세 번째 축을 따라 쌓기

PPT 49페이지 · 인쇄 쪽번호 48

In [186]:
# N38 · PPT 49페이지 (인쇄 쪽번호 48) · 기본 예제
arr1 = np.array([1, 2, 3])
arr2 = np.array([4, 5, 6])
arr = np.dstack((arr1, arr2))
print(arr)
print("shape:", arr.shape)

[[[1 4]
  [2 5]
  [3 6]]]
shape: (1, 3, 2)


**바꿔 보면? 2차원 배열 두 개를 dstack하면?**

In [192]:
# N38 · PPT 49페이지 (인쇄 쪽번호 48) · 변형 예제
a=np.array([[1,2],[3,4]])
print(a.shape)
b=np.array([[10,20],[30,40]])
print(b.shape)

r=np.dstack((a,b))

print(r,"\nshape :", r.shape)
print(r[:,:,0])

(2, 2)
(2, 2)
[[[ 1 10]
  [ 2 20]]

 [[ 3 30]
  [ 4 40]]] 
shape : (2, 2, 2)
[[1 2]
 [3 4]]


(2,2,2)가 되고 마지막 축의 0번째를 선택하면 첫 입력 배열이 나옵니다.

<a id="n39"></a>
## N39 · 배열 연결 예제: 같은 행에 열 추가

PPT 50페이지 · 인쇄 쪽번호 49

In [198]:
# N39 · PPT 50페이지 (인쇄 쪽번호 49) · 기본 예제
a = np.arange(10).reshape(2, -1)
print(a)
b = np.repeat(1, 10).reshape(2, -1)
print(b)
print(np.concatenate([a, b], axis=1))
print(np.hstack([a, b]))

[[0 1 2 3 4]
 [5 6 7 8 9]]
[[1 1 1 1 1]
 [1 1 1 1 1]]
[[0 1 2 3 4 1 1 1 1 1]
 [5 6 7 8 9 1 1 1 1 1]]
[[0 1 2 3 4 1 1 1 1 1]
 [5 6 7 8 9 1 1 1 1 1]]


**바꿔 보면? 가로 연결 대신 세로로 연결하면?**

In [200]:
# N39 · PPT 50페이지 (인쇄 쪽번호 49) · 변형 예제
a=np.arange(10).reshape(2,-1);
b=np.ones((2,5),dtype=int)
print(np.vstack((a,b)))
print(np.hstack((a,b)).shape,np.vstack((a,b)).shape)

[[0 1 2 3 4]
 [5 6 7 8 9]
 [1 1 1 1 1]
 [1 1 1 1 1]]
(2, 10) (4, 5)


hstack은 (2,10), vstack은 (4,5)입니다. 같은 원소 수라도 연결 방향이 다릅니다.

<a id="q02"></a>
### Q02 · 확인2 · PPT 50페이지

아래 두 배열을 가로로 이어 shape (2, 10)으로 만드세요. concatenate와 hstack 두 방법을 비교하세요.

`answer`에 결과를 저장하세요. [풀이](#q02-solution)

In [ ]:
# Q02 · PPT 50페이지 (인쇄 쪽번호 49) · 직접 풀기
a = np.arange(10).reshape(2, -1)
b = np.repeat(1, 10).reshape(2, -1)
answer = None
# 여기에 풀이를 작성하세요.

<a id="n40"></a>
## N40 · array_split(): 배열을 여러 부분으로 나누기

PPT 51페이지 · 인쇄 쪽번호 50

In [209]:
# N40 · PPT 51페이지 (인쇄 쪽번호 50) · 기본 예제
arr = np.array([1, 2, 3, 4, 5, 6])
newarr = np.array_split(arr, 3)
print(newarr)
print(newarr[0])
print(newarr[1])
print(newarr[2])

[array([1, 2]), array([3, 4]), array([5, 6])]
[1 2]
[3 4]
[5 6]


**바꿔 보면? 나누는 개수가 원소 수보다 많으면?**

In [210]:
# N40 · PPT 51페이지 (인쇄 쪽번호 50) · 변형 예제
print(np.array_split(np.arange(7),3))
print(np.array_split(np.arange(3),5))

[array([0, 1, 2]), array([3, 4]), array([5, 6])]
[array([0]), array([1]), array([2]), array([], dtype=int64), array([], dtype=int64)]


7개를 나누면 길이는 3,2,2입니다. 3개를 5개 배열로 나누면 뒤에 빈 배열 둘이 생깁니다.

<a id="n41"></a>
## N41 · copy(): 원본과 독립적인 배열 복사

PPT 52페이지 · 인쇄 쪽번호 51

In [211]:
# N41 · PPT 52페이지 (인쇄 쪽번호 51) · 기본 예제
arr = np.array([1, 2, 3, 4, 5])
x = arr.copy()
arr[0] = 42
print(arr)
print(x)

[42  2  3  4  5]
[1 2 3 4 5]


**바꿔 보면? copy 대신 단순 대입을 하면?**

In [214]:
# N41 · PPT 52페이지 (인쇄 쪽번호 51) · 변형 예제
a=np.array([1,2,3]); b=a
b[0]=99
print(a)

a=np.array([1,2,3]); b=a.copy()
b[0]=99
print(a)

[99  2  3]
[1 2 3]


단순 대입은 같은 배열을 가리키고 copy는 독립된 데이터를 만듭니다.

<a id="n42"></a>
## N42 · where(): 조건에 맞는 원소의 위치 찾기

PPT 53페이지 · 인쇄 쪽번호 52

In [215]:
# N42 · PPT 53페이지 (인쇄 쪽번호 52) · 기본 예제
arr = np.array([1, 2, 3, 4, 5, 4, 4])
x = np.where(arr == 4)
print(x)
print("위치에 해당하는 값:", arr[x])

(array([3, 5, 6]),)
위치에 해당하는 값: [4 4 4]


**바꿔 보면? 일치하는 값이 없거나 4보다 큰 값을 찾으면?**

In [216]:
# N42 · PPT 53페이지 (인쇄 쪽번호 52) · 변형 예제
a=np.array([1,2,3,4,5,4,4])
print(np.where(a==9))
print(np.where(a>4))

(array([], dtype=int64),)
(array([4]),)


없는 값은 빈 인덱스 배열, 4보다 큰 값은 값 5의 위치인 인덱스 4를 반환합니다.

<a id="q03"></a>
### Q03 · 확인3 · PPT 53페이지

두 배열의 같은 위치에서 값이 일치하는 인덱스를 구하세요. 목표 인덱스: [1, 3, 5, 7].

`answer`에 결과를 저장하세요. [풀이](#q03-solution)

In [219]:
# Q03 · PPT 53페이지 (인쇄 쪽번호 52) · 직접 풀기
a = np.array([1, 2, 3, 2, 3, 4, 3, 4, 5, 6])
b = np.array([7, 2, 10, 2, 7, 4, 9, 4, 9, 8])
answer = np.where(a==b)
print(answer)
print(answer[0])
# 여기에 풀이를 작성하세요.

(array([1, 3, 5, 7]),)
[1 3 5 7]


<a id="n43"></a>
## N43 · 여러 조건을 조합하여 검색

PPT 54페이지 · 인쇄 쪽번호 53

In [220]:
# N43 · PPT 54페이지 (인쇄 쪽번호 53) · 기본 예제
a = np.arange(15)
index = np.where((a >= 5) & (a <= 10))
print(a[index])

[ 5  6  7  8  9 10]


**바꿔 보면? AND 대신 OR를 쓰면?**

In [ ]:
# N43 · PPT 54페이지 (인쇄 쪽번호 53) · 변형 예제
a=np.arange(15)
print(a[(a<3)|(a>11)])
print(a[(a>5)&(a<10)])

[ 0  1  2 12 13 14]
[6 7 8 9]


|는 어느 한 조건만 맞아도 선택합니다. 두 번째는 경계를 제외해 6,7,8,9만 선택합니다.

<a id="n44"></a>
## N44 · intersect1d(): 두 배열의 교집합 구하기

PPT 55페이지 · 인쇄 쪽번호 54

In [221]:
# N44 · PPT 55페이지 (인쇄 쪽번호 54) · 기본 예제
a = np.array([1, 2, 3, 2, 3, 4, 3, 4, 5, 6])
b = np.array([7, 2, 10, 2, 7, 4, 9, 4, 9, 8])
print(np.intersect1d(a, b))
print("같은 위치에서 값이 일치하는 인덱스:", np.where(a == b))

[2 4]
같은 위치에서 값이 일치하는 인덱스: (array([1, 3, 5, 7]),)


**바꿔 보면? 중복이 많거나 공통 값이 없으면?**

In [222]:
# N44 · PPT 55페이지 (인쇄 쪽번호 54) · 변형 예제
print(np.intersect1d([1,1,2,3],[1,1,3,3]))
print(np.intersect1d([1,2],[3,4]))

[1 3]
[]


공통 값은 중복 없이 [1,3]으로 정리됩니다. 공통 값이 없으면 빈 배열입니다.

<a id="n45"></a>
## N45 · sort(): 배열의 값 정렬

PPT 56페이지 · 인쇄 쪽번호 55

In [225]:
# N45 · PPT 56페이지 (인쇄 쪽번호 55) · 기본 예제
arr = np.array([3, 2, 0, 1])
print(np.sort(arr))
print("원본은 그대로:", arr)
arr.sort()
print("arr.sort() 실행 후 원본:", arr)

[0 1 2 3]
원본은 그대로: [3 2 0 1]
arr.sort() 실행 후 원본: [0 1 2 3]


**바꿔 보면? 내림차순으로 정렬하려면?**

In [223]:
# N45 · PPT 56페이지 (인쇄 쪽번호 55) · 변형 예제
a=np.array([3,2,0,1])
print(np.sort(a)[::-1])
print(a)

[3 2 1 0]
[3 2 0 1]


정렬한 새 배열을 뒤집습니다. 원본 a는 그대로입니다.

<a id="n46"></a>
## N46 · 정렬할 축을 선택하기

PPT 57페이지 · 인쇄 쪽번호 56

In [226]:
# N46 · PPT 57페이지 (인쇄 쪽번호 56) · 기본 예제
array = np.array([[3, 7, 1], [10, 3, 2], [5, 6, 7]])
print('원본:', array)
print('전체 원소 정렬:', np.sort(array, axis=None))
print('각 행 안에서 정렬(열 번호를 바꿔가며 정렬):', np.sort(array, axis=1))
print('각 열 안에서 정렬(행 번호를 바꿔가며 정렬):', np.sort(array, axis=0))

원본: [[ 3  7  1]
 [10  3  2]
 [ 5  6  7]]
전체 원소 정렬: [ 1  2  3  3  5  6  7  7 10]
각 행 안에서 정렬: [[ 1  3  7]
 [ 2  3 10]
 [ 5  6  7]]
각 열 안에서 정렬: [[ 3  3  1]
 [ 5  6  2]
 [10  7  7]]


기본값 axis=-1은 마지막 축입니다. 2차원에서는 각 행 안에서 정렬하는 axis=1과 같습니다.

<a id="n47"></a>
## N47 · argsort(): 정렬 순서에 해당하는 인덱스 구하기

PPT 58페이지 · 인쇄 쪽번호 57

In [231]:
# N47 · PPT 58페이지 (인쇄 쪽번호 57) · 기본 예제
array = np.array([28, 13, 45, 12, 4, 8, 0])
print(array)
print(array.shape)
print(np.argsort(array))
print("인덱스를 적용한 결과:", array[np.argsort(array)])

[28 13 45 12  4  8  0]
(7,)
[6 4 5 3 1 0 2]
인덱스를 적용한 결과: [ 0  4  8 12 13 28 45]


<a id="n48"></a>
## N48 · unique(): 중복을 제거한 값 구하기

PPT 59페이지 · 인쇄 쪽번호 58

In [232]:
# N48 · PPT 59페이지 (인쇄 쪽번호 58) · 기본 예제
a = np.array([11, 11, 12, 13, 14, 15, 16, 17, 12, 13, 11, 14, 18, 19, 20])
unique_values = np.unique(a)
print(unique_values)

[11 12 13 14 15 16 17 18 19 20]


**바꿔 보면? 각 값이 몇 번 나왔는지도 알려면?**

In [233]:
# N48 · PPT 59페이지 (인쇄 쪽번호 58) · 변형 예제
a=np.array([3,1,3,2,1,3])
values,counts=np.unique(a,return_counts=True)
print(values,counts)

[1 2 3] [2 1 3]


고유 값 [1,2,3]과 각 등장 횟수 [2,1,3]이 같은 순서로 반환됩니다.

<a id="q04"></a>
### Q04 · 확인4 · PPT 59페이지

2차원 배열을 만들고 전체 원소에서 중복을 제거한 값을 구하세요. 아래 보충 입력의 목표 결과는 [1, 2, 3, 4]입니다.

`answer`에 결과를 저장하세요. [풀이](#q04-solution)

In [ ]:
# Q04 · PPT 59페이지 (인쇄 쪽번호 58) · 직접 풀기
a = np.array([[1, 2, 2], [3, 1, 4]])
answer = None
# 여기에 풀이를 작성하세요.

<a id="n49"></a>
## N49 · 기본 산술 연산: 같은 위치끼리 계산

PPT 61페이지 · 인쇄 쪽번호 60

In [ ]:
# N49 · PPT 61페이지 (인쇄 쪽번호 60) · 기본 예제
data = np.array([1, 2])
ones = np.ones(2, dtype=int)
print(data + ones)

[2 3]


**바꿔 보면? 배열 대신 1을 더하거나 길이가 다르면?**

In [ ]:
# N49 · PPT 61페이지 (인쇄 쪽번호 60) · 변형 예제
a=np.array([1,2])
print(a+1)
try:
    print(a+np.array([1,2,3]))
except ValueError as e:
    print("예상한 오류:",e)

[2 3]
예상한 오류: operands could not be broadcast together with shapes (2,) (3,) 


스칼라는 각 원소에 적용됩니다. 길이 2와 3인 두 배열은 브로드캐스팅할 수 없습니다.

<a id="n50"></a>
## N50 · 기본 산술 연산: 원소별 연산

PPT 62페이지 · 인쇄 쪽번호 61

In [ ]:
# N50 · PPT 62페이지 (인쇄 쪽번호 61) · 기본 예제
data = np.array([1, 2])
ones = np.ones(2, dtype=int)
print(data - ones)
print(data * data)
print(data / data)

[0 1]
[1 4]
[1. 1.]


**바꿔 보면? /와 //의 결과는?**

In [ ]:
# N50 · PPT 62페이지 (인쇄 쪽번호 61) · 변형 예제
a=np.array([3,5])
print(a/2)
print(a//2)

[1.5 2.5]
[1 2]


/는 실수 나눗셈, //는 내림 나눗셈의 몫입니다.

<a id="n51"></a>
## N51 · Broadcasting(브로드캐스팅)

PPT 63페이지 · 인쇄 쪽번호 62

In [ ]:
# N51 · PPT 63페이지 (인쇄 쪽번호 62) · 기본 예제
data = np.array([1.0, 2.0])
print(data * 1.6)

[1.6 3.2]


**바꿔 보면? 각 열에 서로 다른 값을 더하면?**

In [ ]:
# N51 · PPT 63페이지 (인쇄 쪽번호 62) · 변형 예제
a=np.array([[1,2,3],[4,5,6]])
print(a+np.array([10,20,30]))
print(a+np.array([[10],[20]]))

[[11 22 33]
 [14 25 36]]
[[11 12 13]
 [24 25 26]]


(3,)은 각 행의 열에, (2,1)은 각 행 전체에 각각의 값을 더하도록 확장됩니다.

<a id="n52"></a>
## N52 · 배열을 하나의 값으로 요약하기

PPT 64페이지 · 인쇄 쪽번호 63

In [ ]:
# N52 · PPT 64페이지 (인쇄 쪽번호 63) · 기본 예제
data = np.array([1.0, 2.0])
print(data.max())
print(data.min())
print(data.sum())
# 아래 그림은 입력이 [1, 2, 3]인 별도 예제입니다.
data = np.array([1, 2, 3])
print('그림의 max, min, sum:', data.max(), data.min(), data.sum())

2.0
1.0
3.0
그림의 max, min, sum: 3 1 6


**바꿔 보면? 2차원 배열에서 축별 합계를 구하면?**

In [ ]:
# N52 · PPT 64페이지 (인쇄 쪽번호 63) · 변형 예제
a=np.array([[1,2,3],[4,5,6]])
print(a.sum(axis=0))
print(a.sum(axis=1))

[5 7 9]
[ 6 15]


axis=0을 줄이면 열별 합 [5,7,9], axis=1을 줄이면 행별 합 [6,15]가 됩니다.

<a id="n53"></a>
## N53 · 자주 사용하는 수학·통계 함수

PPT 65페이지 · 인쇄 쪽번호 64

In [ ]:
# N53 · PPT 65페이지 (인쇄 쪽번호 64) · 기본 예제
# 슬라이드의 모든 함수를 실행 가능한 작은 예제로 확인합니다.
arr1 = np.array([7, 10])
arr2 = np.array([3, 4])
print('덧셈:', np.add(arr1, arr2))
print('뺄셈:', np.subtract(arr1, arr2))
print('원소별 곱셈:', np.multiply(arr1, arr2))
print('나눗셈:', np.divide(arr1, arr2))
print('거듭제곱:', np.power(arr1, arr2))
print('부호 반전:', np.negative(arr1))
print('나머지:', np.mod(arr1, arr2))
print('같은 나머지 연산:', np.remainder(arr1, arr2))
quotient, remainder = np.divmod(arr1, arr2)
print('몫:', quotient, '나머지:', remainder)
print('절댓값:', np.absolute([-3, 0, 2]))
arr = np.array([1, 2, 100])
print('산술평균:', np.mean(arr))
print('중앙값:', np.median(arr))
print('표준편차(ddof=0):', np.std(arr))
print('1차원 내적:', np.dot([1, 2], [3, 4]))
A = np.array([[1, 2], [3, 4]])
print('2차원 행렬 곱:', np.dot(A, np.eye(2)))

덧셈: [10 14]
뺄셈: [4 6]
원소별 곱셈: [21 40]
나눗셈: [2.333333 2.5     ]
거듭제곱: [  343 10000]
부호 반전: [ -7 -10]
나머지: [1 2]
같은 나머지 연산: [1 2]
몫: [2 2] 나머지: [1 2]
절댓값: [3 0 2]
산술평균: 34.333333333333336
중앙값: 2.0
표준편차(ddof=0): 46.435139950496776
1차원 내적: 11
2차원 행렬 곱: [[1. 2.]
 [3. 4.]]


**바꿔 보면? 큰 값 하나가 평균·중앙값에 미치는 영향은?**

In [ ]:
# N53 · PPT 65페이지 (인쇄 쪽번호 64) · 변형 예제
for a in (np.array([1,2,3]),np.array([1,2,100])):
    print("평균:",np.mean(a),"중앙값:",np.median(a),"표준편차:",np.std(a))

평균: 2.0 중앙값: 2.0 표준편차: 0.816496580927726
평균: 34.333333333333336 중앙값: 2.0 표준편차: 46.435139950496776


100은 평균과 표준편차를 크게 높이지만 가운데 순서의 값인 중앙값 2는 유지됩니다.

<a id="n54"></a>
## N54 · 소수점 처리: 버림과 반올림

PPT 66페이지 · 인쇄 쪽번호 65

In [ ]:
# N54 · PPT 66페이지 (인쇄 쪽번호 65) · 기본 예제
print('소수부 버림:', np.trunc([-3.1666, 3.6667]))
print('fix도 0 방향 버림:', np.fix([-3.1666, 3.6667]))
print('소수점 아래 둘째 자리까지:', np.around(3.1666, 2))
print('정확히 중간이면 짝수 쪽:', np.around([2.5, 3.5]))

소수부 버림: [-3.  3.]
fix도 0 방향 버림: [-3.  3.]
소수점 아래 둘째 자리까지: 3.17
정확히 중간이면 짝수 쪽: [2. 4.]


**바꿔 보면? 반올림 자릿수가 0 또는 -1이면?**

In [ ]:
# N54 · PPT 66페이지 (인쇄 쪽번호 65) · 변형 예제
print(np.around([2.5,3.5],decimals=0))
print(np.around([125,135],decimals=-1))

[2. 4.]
[120 140]


0은 정수 자리, -1은 십의 자리로 반올림합니다. 정확히 중간이면 남기는 숫자가 짝수가 되므로 [2,4], [120,140]입니다.

<a id="n55"></a>
## N55 · 소수점 처리: 내림과 올림

PPT 67페이지 · 인쇄 쪽번호 66

In [ ]:
# N55 · PPT 67페이지 (인쇄 쪽번호 66) · 기본 예제
print('내림:', np.floor([-3.1666, 3.6667]))
print('올림:', np.ceil([-3.1666, 3.6667]))
print('10개를 3개씩 담는 데 필요한 상자 수:', int(np.ceil(10 / 3)))

내림: [-4.  3.]
올림: [-3.  4.]
10개를 3개씩 담는 데 필요한 상자 수: 4


**바꿔 보면? 정수와 음수를 올림·내림하면?**

In [ ]:
# N55 · PPT 67페이지 (인쇄 쪽번호 66) · 변형 예제
a=np.array([-2.1,-2.0,2.0,2.1])
print("floor:",np.floor(a))
print("ceil:",np.ceil(a))

floor: [-3. -2.  2.  2.]
ceil: [-2. -2.  2.  3.]


정수는 바뀌지 않습니다. -2.1은 내리면 -3, 올리면 -2입니다.

<a id="n56"></a>
## N56 · Logarithm(로그): 거듭제곱의 지수 구하기

PPT 68페이지 · 인쇄 쪽번호 67

In [ ]:
# N56 · PPT 68페이지 (인쇄 쪽번호 67) · 기본 예제
print('밑이 2인 로그:', np.log2(8))
print('밑이 10인 로그:', np.log10(1000))
print('자연로그:', np.log(np.e))
print('밑이 3인 81의 로그:', np.log(81) / np.log(3))
# frompyfunc는 사용자 함수를 ufunc로 감싸는 별도 도구입니다.
# 아래는 작동 확인용 예이며, 내장 NumPy 연산보다 빨라지는 도구가 아닙니다.
def add_one(x):
    return x + 1
add_one_ufunc = np.frompyfunc(add_one, 1, 1)
result = add_one_ufunc(np.array([1, 2, 3]))
print('사용자 함수 적용:', result, 'dtype:', result.dtype)

밑이 2인 로그: 3.0
밑이 10인 로그: 3.0
자연로그: 1.0
밑이 3인 81의 로그: 4.0
사용자 함수 적용: [2 3 4] dtype: object


**바꿔 보면? 배열 전체에 로그를 적용하면?**

In [ ]:
# N56 · PPT 68페이지 (인쇄 쪽번호 67) · 변형 예제
print(np.log2([1,2,4,8]))
print(np.log10([0.1,1,10,100]))

[0. 1. 2. 3.]
[-1.  0.  1.  2.]


함수가 각 원소에 적용됩니다. 0과 1 사이의 양수는 밑이 1보다 큰 로그에서 음수가 됩니다.

# 종합 실습

PPT 70~76페이지의 실습 1~10입니다.

<a id="e01"></a>
### E01 · 실습1 · PPT 70페이지

모든 원소가 55인 3×3 배열을 만드세요. 힌트: np.full().

`answer`에 결과를 저장하세요. [풀이](#e01-solution)

In [ ]:
# E01 · PPT 70페이지 (인쇄 쪽번호 69) · 직접 풀기

answer = None
# 여기에 풀이를 작성하세요.
#full 함수의 파라미터로 3*3, 55를 넘겨 배열 생성
answer = np.full(shape = (3,3), fill_value=55)
print(answer)

[[55 55 55]
 [55 55 55]
 [55 55 55]]


<a id="e02"></a>
### E02 · 실습2 · PPT 71페이지

원소가 10개인 영벡터를 만들고 다섯 번째 값만 1로 바꾸세요. 목표: [0, 0, 0, 0, 1, 0, 0, 0, 0, 0].

`answer`에 결과를 저장하세요. [풀이](#e02-solution)

In [ ]:
# E02 · PPT 71페이지 (인쇄 쪽번호 70) · 직접 풀기

answer = None
# 여기에 풀이를 작성하세요.
#길이가 10인 영벡터 생성
answer = np.zeros(10)
# 5번째 값을 1로 변경
answer[4] = 1
print(answer)


[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]


<a id="e03"></a>
### E03 · 실습3 · PPT 71페이지

0부터 9까지의 1차원 배열을 역순으로 바꾸세요. 목표: [9, 8, 7, 6, 5, 4, 3, 2, 1, 0].

`answer`에 결과를 저장하세요. [풀이](#e03-solution)

In [ ]:
# E03 · PPT 71페이지 (인쇄 쪽번호 70) · 직접 풀기
a = np.arange(10)
answer = None
# 여기에 풀이를 작성하세요.
#a를 맨 뒤(-1)에서 부터 끝까지(void) 거꾸로(-1) -1:void:-1
answer = a[-1::-1]
print(answer)

[9 8 7 6 5 4 3 2 1 0]


<a id="e04"></a>
### E04 · 실습4 · PPT 71페이지

0부터 8까지의 값을 행 순서로 배치한 3×3 배열을 만드세요. 목표 첫 행은 [0, 1, 2], 마지막 행은 [6, 7, 8]입니다.

`answer`에 결과를 저장하세요. [풀이](#e04-solution)

In [11]:
# E04 · PPT 71페이지 (인쇄 쪽번호 70) · 직접 풀기

answer = None
# 여기에 풀이를 작성하세요.
#arange로 생성한 0~8 벡터를 reshape하여 3*3 shape으로 변경
answer = np.arange(9).reshape(3,3)
print(answer)

[[0 1 2]
 [3 4 5]
 [6 7 8]]


<a id="e05"></a>
### E05 · 실습5 · PPT 72페이지

모든 행이 [0, 1, 2, 3, 4]인 5×5 배열을 만드세요. 배열 생성 함수는 np.zeros와 np.arange만 사용하세요.

`answer`에 결과를 저장하세요. [풀이](#e05-solution)

In [ ]:
# E05 · PPT 72페이지 (인쇄 쪽번호 71) · 직접 풀기

answer = None
# 여기에 풀이를 작성하세요.
# arange로 생성한 0~4 벡터를 * operator로 5개로 만들고 vstack으로 세로로 이어붙임.
answer = np.vstack([np.arange(5)]*5)
print(answer)

[[0 1 2 3 4]
 [0 1 2 3 4]
 [0 1 2 3 4]
 [0 1 2 3 4]
 [0 1 2 3 4]]


<a id="e06"></a>
### E06 · 실습6 · PPT 73페이지

shape (2, 3, 4)인 정수 영배열을 만들고 각 묶음의 두 번째 행만 모두 2로 바꾸세요. 각 묶음의 행은 [0,0,0,0], [2,2,2,2], [0,0,0,0]입니다.

`answer`에 결과를 저장하세요. [풀이](#e06-solution)

In [16]:
# E06 · PPT 73페이지 (인쇄 쪽번호 72) · 직접 풀기

answer = None
# 여기에 풀이를 작성하세요.
#zeros 함수로 shape(2,3,4) zero tensor 생성. dtype으로 element의 data type을 integer로 설정
tensor = np.zeros((2,3,4),dtype = 'i')
#모든 page(:), row index는 1,모든 col(:)을 만족시키는 위치의 값을 2로 변경
tensor[:,1,:] = 2
print(tensor)

[[[0 0 0 0]
  [2 2 2 2]
  [0 0 0 0]]

 [[0 0 0 0]
  [2 2 2 2]
  [0 0 0 0]]]


<a id="e07"></a>
### E07 · 실습7 · PPT 74페이지

3×3 배열 아래에 newRow를 붙여 4×3 배열을 만드세요. 원소별 덧셈이 아니라 행 추가입니다.

`answer`에 결과를 저장하세요. [풀이](#e07-solution)

In [ ]:
# E07 · PPT 74페이지 (인쇄 쪽번호 73) · 직접 풀기
array = np.array([[3, 2, 8], [4, 12, 34], [23, 12, 67]])
newRow = np.array([2, 1, 8])
answer = None
# 여기에 풀이를 작성하세요.
#vstack으로 세로로 이어붙임
answer = np.vstack((array,newRow))
print(answer)

[[ 3  2  8]
 [ 4 12 34]
 [23 12 67]
 [ 2  1  8]]


<a id="e08"></a>
### E08 · 실습8 · PPT 75페이지

((A+B)*(-A/2))를 원소별로 계산하세요. add, negative, divide, multiply 함수를 사용하세요. 목표: [-1.5, -1.5, -1.5].

`answer`에 결과를 저장하세요. [풀이](#e08-solution)

In [27]:
# E08 · PPT 75페이지 (인쇄 쪽번호 74) · 직접 풀기
A = np.ones(3) * 1
B = np.ones(3) * 2
answer = None
# 여기에 풀이를 작성하세요.
#식을 모두 함수로 풀어서 작성
answer = np.multiply(np.add(A,B),np.divide(np.multiply(np.ones(3)*(-1),A),np.ones(3)*2))
#요구 사항을 만족시키기 위해 tolist로 python list 형태로 변경
answer = answer.tolist()
print(answer)

[-1.5, -1.5, -1.5]


<a id="e09"></a>
### E09 · 실습9 · PPT 75페이지

리스트 안에 있는 각 배열의 산술평균을 하나씩 구하세요. 목표: [5.5, 30.6, 34.0].

`answer`에 결과를 저장하세요. [풀이](#e09-solution)

In [28]:
# E09 · PPT 75페이지 (인쇄 쪽번호 74) · 직접 풀기
arrays = [np.array([3, 2, 8, 9]),
          np.array([4, 12, 34, 25, 78]),
          np.array([23, 12, 67])]
answer = None
# 여기에 풀이를 작성하세요.
#arrays list에서 한줄씩 array를 꺼내서 mean을 구한 후 item method를 통해 파이썬 float data type으로 변환함
answer = [np.mean(array).item() for array in arrays]
print(answer)

[5.5, 30.6, 34.0]


<a id="e10"></a>
### E10 · 실습10 · PPT 76페이지

선택 문제: 0과 1이 번갈아 나오는 n×n 체커보드 배열을 만드세요. n=8, 왼쪽 위는 0이며 첫 행은 [0,1,0,1,0,1,0,1]입니다. n=5도 확인하세요.

`answer`에 결과를 저장하세요. [풀이](#e10-solution)

In [ ]:
# E10 · PPT 76페이지 (인쇄 쪽번호 75) · 직접 풀기
n = 8
answer = None
# 여기에 풀이를 작성하세요.
# index가 짝수면 0, 홀수면 1인 n*n 크기의 list 생성
l = [1 if i & 1 else 0 for i in range(n * n)]
# array로 먼저 변환 시키고 n*n shape으로 변환
answer = np.array(l).reshape((n,n))
print("n = 8 :")
print(answer)
print("\n"*2)
#n이 5일 때
n = 5
l = [1 if i & 1 else 0 for i in range(n * n)]
answer = np.array(l).reshape((n,n))
print("n = 5 :")
print(answer)
